File: okj/rendszeruzemelteto/fifa_vilagranglista/rust/FifaVilagRanglista.rs
use std::fs::File;
use std::io::{ BufReader, BufRead };

use encoding_rs::WINDOWS_1252;
use encoding_rs_io::DecodeReaderBytesBuilder;
use itertools::Itertools;

struct Eredmeny {
    csapat: String,
    helyezes: i32,
    valtozas: i32,
    pontszam: i32
}

fn create_eredmeny(line: &String) -> Eredmeny {
    let split = line.split(';').collect::<Vec<&str>>();

    Eredmeny {
        csapat: split[0].to_string(),
        helyezes: split[1].parse::<i32>().unwrap(),
        valtozas: split[2].parse::<i32>().unwrap(),
        pontszam: split[3].parse::<i32>().unwrap()
    }
}

fn main() {
    let input_file = File::open("fifa.txt").unwrap();
    let eredmenyek = BufReader::new(DecodeReaderBytesBuilder::new().encoding(Some(WINDOWS_1252)).build(input_file))
                              .lines()
                              .skip(1)
                              .map(|k| create_eredmeny(&k.unwrap()))
                              .collect::<Vec<Eredmeny>>();

    println!("3. Feladat: Csapatok száma: {}", eredmenyek.len());

    let pontszamok = eredmenyek.iter()
                               .map(|k| k.pontszam as f32)
                               .collect::<Vec<f32>>();

    println!("4. Feladat: Átlagpontszám: {:.2}", pontszamok.iter().sum::<f32>() / pontszamok.len() as f32);

    match eredmenyek.iter().max_by_key(|k| k.valtozas) {
        Some(k) => println!("5. Feladat: Legtöbb helyet javító csapat: {}, helyezés: {}, pontszám: {}", k.csapat, k.helyezes, k.pontszam),
        None => {}
    }

    let volt_e_mo = eredmenyek.iter().any(|k| k.csapat == "Magyarország");

    println!("{}", if volt_e_mo { "6. Feladat: Csapatok között van Magyarország" } else { "6. Feladat: Csapatok között nincs Magyarország" });
    println!("7. Feladat:");

    eredmenyek.iter()
              .sorted_by_key(|k| k.valtozas)
              .group_by(|k| k.valtozas)
              .into_iter()
              .map(|(valtozas, items)| (valtozas, items.count()))
              .filter(|(_, count)| *count > 1)
              .for_each(|(valtozas, count)| println!("    {} helyet változott: {} csapat", valtozas, count));
}

File: okj/rendszeruzemelteto/fifa_vilagranglista/python/FifaVilagRanglista.py
from statistics import mean
from collections import Counter

class Eredmeny:
    def __init__(self, line: str):
        split = line.split(';')

        self.csapat = split[0]
        self.helyezes = int(split[1])
        self.valtozas = int(split[2])
        self.pontszam = int(split[3])


with open('fifa.txt') as file:
    eredmenyek = [ Eredmeny(k) for k in file.readlines()[1:] ]

print(f'3. Feladat: Csapatok száma: {len(eredmenyek)}')

atlagPontszam = mean(k.pontszam for k in eredmenyek)
print('4. Feladat: Átlagpontszám: %.2f' % atlagPontszam)

legtobbetJavitoCsapat = max(eredmenyek, key = lambda k: k.valtozas)
print(f'5. Feladat: Legtöbbet javító csapat: {legtobbetJavitoCsapat.csapat}, helyezés: {legtobbetJavitoCsapat.helyezes}, pontszam: {legtobbetJavitoCsapat.pontszam}')

voltEMo = any(True for k in eredmenyek if k.csapat == 'Magyarország')
print('6. Feladat: Csapatok között van Magyarország' if voltEMo else '6. Feladat: Csapatok között nincs Magyarország')
print('7. Feladat:')

for valtozas, db in Counter(k.valtozas for k in eredmenyek).items():
    if db > 1:
        print(f'    {valtozas} helyet változott: {db} csapat')